manticore/utils/event.py

```diff
@@ -1,168 +1,77 @@
-        else:
-            method = self.emit
-        if len(self):
-            signal.connect(method)
-        else:
-            self._forwards.setdefault(signal, set()).add(method)
-
-    def __iadd__(self, dest):
-        self.connect(dest)
-        return self
-
-    def disconnect(self, dest):
-        try:
-            if inspect.ismethod(dest):
-                obj, impl = dest.__self__, dest.__func__
-                self._methods[obj].remove(impl)
+    def __setstate__(self, state):
+        ''' It wont get serialized by design, user is responsible to reconnect'''
+        self._signals = dict()
+        self._forwards = WeakKeyDictionary()
+        return True
+
+    def __getstate__(self):
+        return {}
+
+    def _unref(self, robj):
+        # this is called when an object that has subscribed to events emitted 
+        # here has recently been garbage collected
+        # This simply removes all callback methods associated with that object
+        # Also if no more callbacks at all for an event name it deletes the event entry
+        remove = set()
+        for name, bucket in self._signals.items():
+            if robj in bucket:
+                del bucket[robj]
+            if len(bucket) == 0:
+                remove.add(name)
+        for name in remove:
+            del self._signals[name]
+
+    def _get_signal_bucket(self, name):
+        #Each event name has a bucket of callback methods
+        #A bucket is a dictionary obj -> set(method1, method2...)
+        return self._signals.setdefault(name,  dict())
+
+    def publish(self, name, *args, **kwargs):   
+        bucket = self._get_signal_bucket(name)
+        for obj, methods in bucket.items():
+            for callback in methods:
+                callback(obj(), *args, **kwargs)
+
+        #The include_source flag indicates to prepend the source of the event in
+        # the callback signature. This is set on forward_events_from/to 
+        for sink, include_source in self._forwards.items():
+            if include_source:
+                sink.publish(name, self, *args, **kwargs)
             else:
-                self._functions.remove(dest)
-        except KeyError:
-            raise SignalDisconnectedError()
-
-    def __isub__(self, dest):
-        self.disconnect(dest)
-        return self
-
-    def reset(self):
-        self._functions.clear()
-        self._methods.clear()
-        self._forwards.clear()
-
+                sink.publish(name, *args, **kwargs)
+
+    def subscribe(self, name, method):
+        assert inspect.ismethod(method)
+        obj, callback = method.__self__, method.__func__
+        bucket = self._get_signal_bucket(name)
+        robj = ref(obj, self._unref)  #see unref() for explanation
+        bucket.setdefault(robj, set()).add(callback)
+
+    def forward_events_from(self, source, include_source=False):
+        if isinstance(source, Eventful):
```

Comment:
i think these functions should have corresponding error handling, i think they should both raise TypeError also if source is not an Eventful